arangod/RocksDBEngine/RocksDBPersistedLog.cpp

```diff
@@ -0,0 +1,293 @@
+        while (wb.Count() < 1000 && nextReqToWrite != std::end(pendingRequests)) {
+          auto* log_ptr =
+              dynamic_cast<RocksDBPersistedLog*>(nextReqToWrite->log.get());
+          TRI_ASSERT(log_ptr != nullptr);
+          if (auto res = log_ptr->prepareWriteBatch(*nextReqToWrite->iter, wb);
+              res.fail()) {
+            return res;
+          }
+          TRI_ASSERT(nextReqToWrite->iter->next() == std::nullopt);
+          ++nextReqToWrite;
+        }
+        {
+          if (auto s = _db->Write({}, &wb); !s.ok()) {
+            return rocksutils::convertStatus(s);
+          }
+
+          if (lane._waitForSync) {
+            if (auto s = _db->SyncWAL(); !s.ok()) {
+              // At this point we have to make sure that every previous log entry is synced
+              // as well. Otherwise we might get holes in the log.
+              return rocksutils::convertStatus(s);
+            }
+          }
+        }
+
+        // resolve all promises in [nextReqToResolve, nextReqToWrite)
+        for (; nextReqToResolve != nextReqToWrite; ++nextReqToResolve) {
+          nextReqToResolve->promise.setValue(TRI_ERROR_NO_ERROR);
+        }
+      }
+
+      return Result{TRI_ERROR_NO_ERROR};
+    });
+
+    // resolve all promises with the result
+    if (result.fail()) {
+      for (; nextReqToResolve != std::end(pendingRequests); ++nextReqToResolve) {
+        // If a promise is fulfilled before (with a value), nextReqToResolve
+        // should always be increased as well; meaning we only exactly iterate
+        // over the unfulfilled promises here.
+        TRI_ASSERT(!nextReqToResolve->promise.isFulfilled());
+        nextReqToResolve->promise.setValue(result);
+      }
+    }
+  }
+}
+
+auto RocksDBLogPersistor::persist(std::shared_ptr<PersistedLog> log,
+                                  std::unique_ptr<PersistedLogIterator> iter,
+                                  WriteOptions const& options) -> futures::Future<Result> {
+  auto p = futures::Promise<Result>{};
+  auto f = p.getFuture();
+
+  Lane& lane = _lanes[options.waitForSync ? 0 : 1];
+  TRI_ASSERT(lane._waitForSync == options.waitForSync);
+
+  {
+    std::unique_lock guard(lane._persistorMutex);
+    lane._pendingPersistRequests.emplace_back(std::move(log), std::move(iter),
+                                              std::move(p));
+    // TODO this code limits the total amount of threads working in the persistor
+    //      maybe we want to limit it for each lane as well?
+
+    // TODO add options for max number of persistor threads for replicated logs
+    size_t num_retries = 0;
+    while (true) {
+      auto lambda = fu2::unique_function<void() noexcept>{
+          [self = shared_from_this(), &lane]() noexcept {
+            self->runPersistorWorker(lane);
+          }};
+
+      if (lane._activePersistorThreads == 0 ||
+          (lane._pendingPersistRequests.size() > 100 && lane._activePersistorThreads < 2)) {
```

Comment:
If this `if` branch is not taken, the loop will be a busy loop, which will only create `lambda` objects in a loop, without any pause.

Reply:
Whoopse, so I think we should do a break in the else case?

Reply:
Or let's just write `while(lane._activePersistorThreads == 0 || (lane._pendingPersistRequests.size() > 100 && lane._activePersistorThreads < 2)) { ...`?

Reply:
We simplified the code. Now a user either decides to start a new thread or not. If it decides to do so, it commits to.